base/not_intersecting_intervals.hpp

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include "base/assert.hpp"
+
+#include <set>
+
+namespace base
+{
+/// \brief Implements data structure, that allows to add intervals with |AddInterval()|
+///        method. This method returns true if all previous intervals, which were added
+///        by |AddInterval()|, don't overlap with current.
+template <typename T>
+class NotIntersectingIntervals
+{
+public:
+  /// \return true if there are no such intervals, that intersect the [left, right] interval.
```

Comment:
Does this method it modify the structure?

Reply:
```cpp
  /// \brief Adds new interval to set if it doesn't intersect with any that have already added.
  /// \return true if there are no such intervals, that intersect the [left, right] interval.
  bool AddInterval(T left, T right);
```